### DISCUSSION TITLE
Accepting an array of `Langchain::Messages::*` instances when calling `Langchain::LLM::*#chat(messages: [...])`

### DISCUSSION THREAD

**Discussion by andreibondarev** [patterns-ai-core/langchainrb#629] [Ideas]
I'm thinking that the LLM#chat() methods would be able to accept the messages: [] array that is a collection of `Langchain::Messages::*` instances. The benefits of this are that people won't need to know how the message hash is structured.
Examples:
```ruby
# OpenAI:
{role:"user", content:"hi"}
# Gemini:
{role:"user", parts: [{text:"hi"}]}
```

Proposed usage:
```ruby

@kokuyouwind I've been thinking that the #chat(messages: []) method could accept the Langchain::Messages::* instances directly. For example:

message_1 = Langchain::Messages::AnthropicMessage.new(role:"user", content:"hi!")
message_2 = Langchain::Messages::AnthropicMessage.new(role:"assistant", content:"Hey! How can I help?")
message_3 = Langchain::Messages::AnthropicMessage.new(role:"assistant", content:"Help me debug my computer")

Langchain::LLM::Anthropic.new(...).chat(messages: [message_1, message_2, message_3])
```

**Comment by andreibondarev** [comment]
@kokuyouwind I created a discussion thread to talk through it here, follow-up to https://github.com/patterns-ai-core/langchainrb/issues/603#issuecomment-2099553633

@drnic I think I recall you had asked something similar?!

**Comment by kokuyouwind** [comment]
@andreibondarev 
(This comment is reprinted from https://github.com/patterns-ai-core/langchainrb/issues/603#issuecomment-2121388673 )

I think it is a very excellent idea.
To make it more generic, I think it would be better to make it not an `AnthropicMessage` but a per-role message instance that is LLM-independent.

```ruby
message_1 = Langchain::Messages::UserMessage.new("hi!")
message_2 = Langchain::Messages::AssistantMessage.new("Hey! How can I help?")
message_3 = Langchain::Messages::AssistantMessage.new("Help me debug my computer")

Langchain::LLM::Anthropic.new(...).chat(messages: [message_1, message_2, message_3])
```

The class names above are aligned with the role notation, but could be aligned with [Python's LangChain Messages](https://python.langchain.com/v0.1/docs/modules/model_io/chat/message_types/), such as `HumanMessage`, `AIMessage`, etc.

**Comment by andreibondarev** [comment]
@kokuyouwind I also looked up what LlamaIndex does. It's got simple [ChatMessage ](https://github.com/run-llama/llama_index/blob/main/llama-index-core/llama_index/core/base/llms/types.py#L29) class. 

I think the main takeaway here is that we need a ChatMessage class, and maybe shortcuts for UserMessage, AssistantMessage, FunctionMessage and that the LLM classes themselves are responsibly for serializing the ChatMessage instance on its own.

**Comment by kokuyouwind** [comment]
@andreibondarev 
Certainly, there is little need for separate classes for each role, and it is a good design to have a common abstraction in the `ChatMessage` class.
Having the client class of each LLM responsible for serializing messages is exactly what I wanted.

**Reply by andreibondarev** [reply]
@kokuyouwind Agreed, that should be the direction to go into.

**Reply by kokuyouwind** [reply]
(I just realized that it is a Discussion, not an Issue, and that I can use Reply...lol)

Are there any other viewpoints that we should discuss? If there is noting to discuss, I will try to create a draft PR.
We could also consider integrating the Messages class under lib/langchain/assistants/messages into ChatMessage. 
If it looks like I should give it a try, because it will probably conflict with the work on https://github.com/patterns-ai-core/langchainrb/issues/543, I would like to coordinate the timing of the work.

**Reply by andreibondarev** [reply]
@kokuyouwind Go for it! 🚀 

We should produce small PRs. I'd rather introduce small incremental changes than large rewrites of functionality.

**Reply by kokuyouwind** [reply]
I see.
I will make a draft with the following contents.

* Introduce `Langchain::Messages::ChatMessages`.
  * Once the function calling response is not considered, only :system, :user, and :assistant are accepted for roles.
* Make each LLM accept `[ChatMessages]` in messages of `chat` method.
  * For compatibility, continue to accept `[Hash]` as well.
  *  To support the above polymorphic behavior, change `Langchain::LLM::Base#chat` to call `chat_by_raw_messages` and `chat_by_chat_messages` separately depending on `messages`, and rename existing `chat` methods of each client to `chat_by_raw_messages` (I think this will be the least friction)

**Reply by andreibondarev** [reply]
I think we can handle it inside of `chat()` methods, right?

Pseudo-code:
```ruby
class OpenAI

  def chat(messages: [], ...)
    if messages.is_a?(Hash)
      chat_by_chat_messages()
    elsif messages.is_a?(Langchain::Messages::ChatMessage)
      chat_by_raw_messages()
```

Maybe the class is really just called `Langchain::ChatMessage`?

**Reply by kokuyouwind** [reply]
> I think we can handle it inside of chat() methods, right?

Yes, I am considering something like that.

In addition, In the way of modifying each LLM client, we would have to write “if statements” and “convert `Langchain::Messages::ChatMessage` to raw hash” for every LLM client, which would duplicate a lot of code.
My idea is to give a good default in `Base` and then override it as needed in each client.
It roughly looks like this:

```ruby
class Langchain::LLM::Base
  def chat(params = {}, ...)
    if params[:messages].all? { _1.is_a?(Hash) }
      chat_by_raw_messages(params, ...)
    elsif params[:messages].all? { _1.is_a?(Langchain::Messages::ChatMessage) }
      chat_by_chat_messages(params, ...)
    else
      # When an array with mixed raw hash and ChatMessage is passed, it is debatable whether to allow this or not, 
      # but I think it is better to make an error
      raise InvalidArgumentError
    end
  end

  def chat_by_chat_messages(params, ...)
    chat_by_raw_messages(
      params.merge(messages: params[:messages].map(&:to_hash),
      ...
    )
  end

  def chat_by_raw_messages(...)
    raise NotImplementedError, "#{self.class.name} does not support chat"
  end
end
```

This allows most models to support `ChatMessage` by simply changing `chat` to `chat_by_raw_messages` while retaining existing behavior.
For clients like `Anthropic` that cannot handle parts of `ChatMessage` directly, we can override `chat_by_chat_messages` to define how to convert to raw messages.

I think this method would have the least impact on the existing code.

**Reply by andreibondarev** [reply]
@kokuyouwind I think the "catch all" in `Langchain::LLM::Base` is a good idea but I'd rather be explicit. You're basically putting the content of these [methods](https://github.com/patterns-ai-core/langchainrb/blob/main/lib/langchain/assistants/messages/openai_message.rb#L43-L50) into `chat_by_chat_messages()`.

I'm thinking maybe a better implementation is something like this:
```ruby
def chat(params = {})
  if params[:messages].all? { _1.is_a?(Langchain::ChatMessage) }
    params[:messages] = params[:messages].map &:to_openai_messages
  end
  ...
  client.chat(**params)
```
What do you think?

**Reply by kokuyouwind** [reply]
Certainly, it might be preferable to write it explicitly if possible.
While I'm a bit concerned about the amount of duplicate code, I'd like to create a draft PR once based on that approach.

**Reply by andreibondarev** [reply]
Let's do it!

**Reply by kokuyouwind** [reply]
I created the first draft in #652.
I only implemented adapters for OpenAI and Anthropic. The implementation for other adapters should be similar to OpenAI.
If there are no issues with this implementation approach, I plan to add support for the other adapters and request a review.

**Reply by andreibondarev** [reply]
@kokuyouwind Thank you, I left some comments.

**Reply by kokuyouwind** [reply]
Thank you, I've replied within the PR.
If there seems to be a need for discussion around the tools, it's also okay to bring it over to this Discussion Thread.
